### ISSUE TITLE
Expenditure of time

### ISSUE THREAD

**Issue by veryhannibal** [facebookresearch/CrypTen#295]
**I trained a simple CNN on 2000 MNIST images with batch size 32 for 30 epochs, it cost about 1 hour. Is it common? 
I just modifed the code of tutorial 7, the code is blow:**

import crypten
import torch
import crypten.mpc as mpc
import crypten.communicator as comm
import torch.nn as nn
import torch.nn.functional as F
import matplotlib.pyplot as plt
import time

crypten.init()
torch.set_num_threads(1)

# Define an example network
class ExampleNet(nn.Module):
    def __init__(self):
        super(ExampleNet, self).__init__()
        self.conv1 = nn.Conv2d(1, 16, kernel_size=5, padding=0)
        self.fc1 = nn.Linear(16 * 12 * 12, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        out = self.conv1(x)
        out = F.relu(out)
        out = F.avg_pool2d(out, 2)
        out = out.view(-1, 16 * 12 * 12)
        out = self.fc1(out)
        out = F.relu(out)
        out = self.fc2(out)
        return out

start = time.time()

crypten.common.serial.register_safe_class(ExampleNet)

# Define source argument values for Alice and Bob
ALICE = 0
BOB = 1

model_plaintext = ExampleNet()
#print(model_plaintext.state_dict())

train_size = 2000
test_size = 200

dummy_input = torch.empty(1, 1, 28, 28)
model = crypten.nn.from_pytorch(model_plaintext, dummy_input)
model.encrypt()

labels = torch.load('tmp/train_labels.pth')[:train_size]
labels = labels.long()
label_eye = torch.eye(10)
labels_one_hot = label_eye[labels]
test_labels = torch.load("tmp/test_labels.pth")[:test_size]
test_label_eye = torch.eye(10)
test_labels_one_hot = test_label_eye[test_labels]

@mpc.run_multiprocess(world_size=2)
def run_encrypted_training():
    global model
    # Load data:
    x_alice_enc = crypten.load_from_party('tmp/alice_train.pth', src=ALICE)
    x_bob_enc = crypten.load_from_party('tmp/bob_train.pth', src=BOB)

    # crypten.print(x_alice_enc.size())
    # crypten.print(x_bob_enc.size())

    # Combine the feature sets: identical to Tutorial 3
    x_combined_enc = crypten.cat([x_alice_enc, x_bob_enc], dim=2)[:train_size]

    # Reshape to match the network architecture
    x_combined_enc = x_combined_enc.unsqueeze(1)
    print(x_combined_enc.size())
    y_enc_one_hot = crypten.cryptensor(labels_one_hot, requires_grad=True)
    x_test_alice_enc = crypten.load_from_party("tmp/alice_test.pth", src=ALICE)
    x_test_bob_enc = crypten.load_from_party("tmp/bob_test.pth", src=BOB)
    x_test_combined_enc = crypten.cat([x_test_alice_enc, x_test_bob_enc], dim=2).unsqueeze(1)[:test_size]
    y_test_enc_one_hot = crypten.cryptensor(test_labels_one_hot)

    # Commenting out due to intermittent failure in PyTorch codebase

    # Initialize a plaintext model and convert to CrypTen model
    #pytorch_model = ExampleNet()
    #model = crypten.nn.from_pytorch(pytorch_model, dummy_input)
    #model.encrypt()
    # # Set train mode

    # Define a loss function
    loss = crypten.nn.CrossEntropyLoss()

    # Define training parameters
    learning_rate = 0.05
    num_epochs = 30
    batch_size = 100
    num_batches = x_combined_enc.size(0) // batch_size

    test_batch_size = 50
    num_test_batches = test_size // test_batch_size

    test_acc_list = []
    rank = comm.get().get_rank()
    for i in range(num_epochs):
        if i + 1 >= 5 and (i + 1) % 5 == 0:
            learning_rate *= 0.9
        crypten.print(f"Epoch {i+1} in progress:")
        model.train()
        for batch in range(num_batches):
            # define the start and end of the training mini-batch
            start, end = batch * batch_size, (batch + 1) * batch_size

            # construct CrypTensors out of training examples / labels
            x_train = x_combined_enc[start:end]
            #y_batch = labels_one_hot[start:end]
            #y_train = crypten.cryptensor(y_batch, requires_grad=True)
            y_train = y_enc_one_hot[start:end]

            # perform forward pass:
            output = model(x_train)
            loss_value = loss(output, y_train)

            # set gradients to "zero"
            model.zero_grad()

            # perform backward pass:
            loss_value.backward()

            # update parameters
            model.update_parameters(learning_rate)

            # Print progress every batch:
            batch_loss = loss_value.get_plain_text()
            crypten.print(f"\tBatch {(batch + 1)} of {num_batches} Loss {batch_loss.item():.4f}")

        crypten.print("--------Evaluation--------")
        model.eval()
        test_running_loss = 0
        test_correct = 0
        for test_batch in range(num_test_batches):
            start, end = test_batch * test_batch_size, (test_batch + 1) * test_batch_size
            x_test = x_test_combined_enc[start:end]
            y_test_true_batch = test_labels[start:end]
            #y_test_batch = test_labels_one_hot[start:end]
            #y_test = crypten.cryptensor(y_test_batch)
            y_test = y_test_enc_one_hot[start:end]
            test_output = model(x_test)
            test_loss_value = loss(test_output, y_test)
            test_batch_loss = test_loss_value.get_plain_text()
            test_running_loss += test_batch_loss.item()
            test_output_plain_text = test_output.get_plain_text().max(1)[1]
            test_correct += test_output_plain_text.eq(y_test_true_batch).sum().item()
        crypten.print("Test accuracy on epoch-{} is {}.".format(i + 1, round(test_correct / test_size, 4)))
        crypten.print("Test loss on epoch-{} is {}.".format(i + 1, round(test_running_loss / num_test_batches, 4)))
        test_acc_list.append(round(test_correct / test_size, 4))

**Comment by knottb** [comment]
First, computation in MPC is always going to be more expensive than plaintext counterparts. The performance gap has decreased dramatically over the past few years.

Regarding your example, it seams reasonable that this kind of training might take this long on CPU. My recollection having trained CNN's on MNIST is that 30 epochs can take some time even in plaintext on CPU. Here are a few things you can do to improve the runtime (even in plaintext):

1. Increase batch size: MPC in particular benefits strongly from increased batching. The `ExampleNet` model you are using can still train will with much larger batch sizes than 32. I typically use 256, but my understanding is this can be increased quite a bit more before seeing artifacts during training.

2. Reduce number of epochs: My understanding is that model training for this example will converge in 3-5 epochs rather than 30. This would reduce your runtime roughly by a factor of 10. My memory could be a bit off here, but I think 3-5 epochs is the "right" amount in this case.

3. Use GPUs: In plaintext, training even small models on CPU can be prohibitively expensive (in terms of time). Scaling to larger models is really only possible with the advent of GPUs. We have added support for use of GPU's in CrypTen, and tend to see performance speedups around 10-100x for large models.

I hope this is helpful.

**Comment by veryhannibal** [comment]
> First, computation in MPC is always going to be more expensive than plaintext counterparts. The performance gap has decreased dramatically over the past few years, but there is still a sizable gap.
> 
> Regarding your example, it seams reasonable that this kind of training might take this long on CPU. My recollection having trained CNN's on MNIST is that 30 epochs can take some time even in plaintext on CPU. Here are a few things you can do to improve the runtime (even in plaintext):
> 
> 1. Increase batch size: MPC in particular benefits strongly from increased batching. The `ExampleNet` model you are using can still train will with much larger batch sizes than 32. I typically use 256, but my understanding is this can be increased quite a bit more before seeing artifacts during training.
> 2. Reduce number of epochs: My understanding is that model training for this example will converge in 3-5 epochs rather than 30. This would reduce your runtime roughly by a factor of 10. My memory could be a bit off here, but I think 3-5 epochs is the "right" amount in this case.
> 3. Use GPUs: In plaintext, training even small models on CPU can be prohibitively expensive (in terms of time). Scaling to larger models is really only possible with the advent of GPUs. We have added support for use of GPU's in CrypTen, and tend to see performance speedups around 10-100x for large models.
> 
> I hope this is helpful.

Thanks a lot, your suggestion is really helpful for me.  Does CrypTen currently supports GPU? I have been using CrypTen for only 2 weeks, The reason I ask this question is that I compared CrypTen with TF-encrypted under the same condition, I found tf-encrypted is much more fast, so I wonder if there is something wrong with my code. I am not going to say tf-encrypted is better, actually I prefer to use CrepTen, because I am familiar with Pytorch and it is really convenient to use torch to build model and encrypt it with Crypten.
Thank you very much!
@knottb

**Comment by lvdmaaten** [comment]
I forget if it is already in the current release on PyPi, but if you install CrypTen from source you should have full GPU support for all functions and models. Depending on your model, this can dramatically speed up CrypTen (we have seen speed-ups of close to 100x for certain models).

**Comment by veryhannibal** [comment]
Thank you very much! I will install from source. So which branch should I used? Branch master?  @lvdmaaten

**Comment by lvdmaaten** [comment]
Yeah `master` should be fine.
